packages/aws-cdk/lib/notices.ts

```diff
@@ -244,21 +260,61 @@ function formatOverview(text: string) {
   return '\t' + heading + content;
 }
 
-function frameworkVersion(outdir: string): string | undefined {
-  const tree = loadTree().tree;
+/**
+ * Whether any component in the tree matches any component in the query.
+ */
+function match(query: Component[], tree: Node): boolean {
+  return some(tree, node => {
+    return query.some(component =>
+      node.constructInfo?.fqn.startsWith(component.name) &&
+      semver.satisfies(node.constructInfo?.version ?? '', component.version));
+  });
+}
 
-  if (tree?.constructInfo?.fqn.startsWith('aws-cdk-lib')
-    || tree?.constructInfo?.fqn.startsWith('@aws-cdk/core')) {
-    return tree.constructInfo.version;
+function loadTree(outdir: string) {
+  try {
+    return fs.readJSONSync(path.join(outdir, 'tree.json'));
+  } catch (e) {
+    debug(`Failed to get tree.json file: ${e}`);
+    return {};
   }
-  return undefined;
+}
 
-  function loadTree() {
-    try {
-      return fs.readJSONSync(path.join(outdir, 'tree.json'));
-    } catch (e) {
-      debug(`Failed to get tree.json file: ${e}`);
-      return {};
+/**
+ * Source information on a construct (class fqn and version)
+ */
+interface ConstructInfo {
+  readonly fqn: string;
+  readonly version: string;
+}
+
+/**
+ * A node in the construct tree.
+ * @internal
+ */
+interface Node {
```

Comment:
`Node` is very minimal, given that we could have a lot of trees going on. `ConstructTreeInfoNode` or something?